### PR TITLE
fix: [CDS-44788]: enforcing time range in cd page workflowExecutions query

### PR DESCRIPTION
**Pull request by lasdox**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/38866)

### CODE REVIEW

File: 400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionOptimizationHelper.java [hunk 1]
```diff
@@ -52,7 +64,91 @@ public void enforceAppIdFromChildrenEntities(PageRequest<WorkflowExecution> page
...
+    if (!appIdFilterOpt.isPresent()) {
+      createdAtList.forEach(filter -> {
+        switch (filter.getOp()) {
+          case GE:
+          case GT:
+            createdAtMap.put("startedAt", Long.parseLong(filter.getFieldValues()[0].toString()));
+            break;
+          case LT:
+          case LT_EQ:
+            createdAtMap.put("endAt", Long.parseLong(filter.getFieldValues()[0].toString()));
```
Comment (abhinav-harness) [inline 400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionOptimizationHelper.java:95]: npe safety needed?

**Comment by abhinav-harness** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java` line 68 at r2](https://reviewable.io/reviews/harness/harness-core/38866#-NEbg2Up2IcOeSTmqmyb:-NEbg2Up2IcOeSTmqmyc:bwax8b4) ([raw file](https://github.com/harness/harness-core/blob/0d6ee1424640e8e9f8ba135e9860fdd8c7c0f864/400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java#L68)):*
> ```Java
> 
>     if (isEmpty(searchFiltersForTime) && appIdFilterOpt.isPresent()) {
>       log.info("Automatically adding search filter of 3 months");
> ```

1 month log



_Code quote:_
```Java
log
```

**Review by abhinav-harness — CHANGES_REQUESTED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38866)* status: 1 change requests. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java` line 85 at r2](https://reviewable.io/reviews/harness/harness-core/38866#-NEbi1D9Alx1z4wUWdTy:-NEbi1DAB1LZ4s87R1mO:ba12j5h) ([raw file](https://github.com/harness/harness-core/blob/0d6ee1424640e8e9f8ba135e9860fdd8c7c0f864/400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java#L85)):*
> ```Java
>           case GE:
>           case GT:
>             createdAtMap.put("startedAt", getTimeValueFromFilter(String.valueOf(filter.getFieldValues()[0])));
> ```

WorkflowExecutionKeys.createdAt
___
*[`400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java` line 89 at r2](https://reviewable.io/reviews/harness/harness-core/38866#-NEbi51yAcoSJCaxL-YP:-NEbi51yAcoSJCaxL-YQ:b-7ropln) ([raw file](https://github.com/harness/harness-core/blob/0d6ee1424640e8e9f8ba135e9860fdd8c7c0f864/400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java#L89)):*
> ```Java
>           case LT:
>           case LT_EQ:
>             createdAtMap.put("endAt", getTimeValueFromFilter(String.valueOf(filter.getFieldValues()[0])));
> ```

ditto
___
*[`400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java` line 110 at r2](https://reviewable.io/reviews/harness/harness-core/38866#-NEbivFTAoruubJiXZDc:-NEbivFTAoruubJiXZDd:bjx5570) ([raw file](https://github.com/harness/harness-core/blob/0d6ee1424640e8e9f8ba135e9860fdd8c7c0f864/400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java#L110)):*
> ```Java
>         Duration duration = Duration.between(startedAt, endAt);
>         if (duration.compareTo(MAXIMUM_DURATION_WITHOUT_APPID) > 0) {
>           startedAt = endAt.minus(MAXIMUM_DURATION_WITHOUT_APPID);
> ```

throw exception as we should not modify user input if he has given both the times
___
*[`400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java` line 132 at r2](https://reviewable.io/reviews/harness/harness-core/38866#-NEbjLOK24CTVERbqy5v:-NEbjLOK24CTVERbqy5w:b4sa3om) ([raw file](https://github.com/harness/harness-core/blob/0d6ee1424640e8e9f8ba135e9860fdd8c7c0f864/400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java#L132)):*
> ```Java
>         endFilterBuilder.fieldName(WorkflowExecutionKeys.createdAt)
>             .fieldValues(new Object[] {startedAt.toEpochMilli()})
>             .op(GE);
> ```

lets do GTE

**Comment by lasdox** [comment]
> > Reviewed 4 of 4 files at r2, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/harness/harness-core/38866)_ status: 1 change requests. 0 of 1 approvals obtained.
> 
> _[`400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java` line 85 at r2](https://reviewable.io/reviews/harness/harness-core/38866#-NEbi1D9Alx1z4wUWdTy:-NEbi1DAB1LZ4s87R1mO:ba12j5h) ([raw file](https://github.com/harness/harness-core/blob/0d6ee1424640e8e9f8ba135e9860fdd8c7c0f864/400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java#L85)):_
> 
> > ```java
> >           case GE:
> >           case GT:
> >             createdAtMap.put("startedAt", getTimeValueFromFilter(String.valueOf(filter.getFieldValues()[0])));
> > ```
> 
> WorkflowExecutionKeys.createdAt

this is not applicable, im populating the map to identify the LTE or GTE values, they need to be different values

> 
> _[`400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java` line 89 at r2](https://reviewable.io/reviews/harness/harness-core/38866#-NEbi51yAcoSJCaxL-YP:-NEbi51yAcoSJCaxL-YQ:b-7ropln) ([raw file](https://github.com/harness/harness-core/blob/0d6ee1424640e8e9f8ba135e9860fdd8c7c0f864/400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java#L89)):_
> 
> > ```java
> >           case LT:
> >           case LT_EQ:
> >             createdAtMap.put("endAt", getTimeValueFromFilter(String.valueOf(filter.getFieldValues()[0])));
> > ```
> 
> ditto

this is not applicable, im populating the map to identify the LTE or GTE values, they need to be different values

> 
> _[`400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java` line 110 at r2](https://reviewable.io/reviews/harness/harness-core/38866#-NEbivFTAoruubJiXZDc:-NEbivFTAoruubJiXZDd:bjx5570) ([raw file](https://github.com/harness/harness-core/blob/0d6ee1424640e8e9f8ba135e9860fdd8c7c0f864/400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java#L110)):_
> 
> > ```java
> >         Duration duration = Duration.between(startedAt, endAt);
> >         if (duration.compareTo(MAXIMUM_DURATION_WITHOUT_APPID) > 0) {
> >           startedAt = endAt.minus(MAXIMUM_DURATION_WITHOUT_APPID);
> > ```
> 
> throw exception as we should not modify user input if he has given both the times

Done

> 
> _[`400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java` line 132 at r2](https://reviewable.io/reviews/harness/harness-core/38866#-NEbjLOK24CTVERbqy5v:-NEbjLOK24CTVERbqy5w:b4sa3om) ([raw file](https://github.com/harness/harness-core/blob/0d6ee1424640e8e9f8ba135e9860fdd8c7c0f864/400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java#L132)):_
> 
> > ```java
> >         endFilterBuilder.fieldName(WorkflowExecutionKeys.createdAt)
> >             .fieldValues(new Object[] {startedAt.toEpochMilli()})
> >             .op(GE);
> > ```
> 
> lets do GTE

GE means GTE in this case, i checked the query result

**Comment by abhinav-harness** [comment]
*[`400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java` line 109 at r3](https://reviewable.io/reviews/harness/harness-core/38866#-NEcCp4V7xp0ifNmxNz5:-NEcCp4V7xp0ifNmxNz6:b-hll4ok) ([raw file](https://github.com/harness/harness-core/blob/a9f914a9805ecef070bf24723be073ef3851345f/400-rest/src/main/java/software/wings/service/impl/WorkflowExecutionTimeFilterHelper.java#L109)):*
> ```Java
>         if (duration.compareTo(MAXIMUM_DURATION_WITHOUT_APPID) > 0) {
>           throw new InvalidRequestException("Maximum time range without appId is 3 months.");
>         }
> ```

where do we check that without appid it is 1 month and throw error?

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38866)* status: :shipit: complete! ✓

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 3 of 6 files at r4, 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/38866)* status: :shipit: complete! ✓

**Pull request merged**